include/nbl/builtin/hlsl/tgmath/impl.hlsl

```diff
@@ -0,0 +1,585 @@
+{
+	static FloatingPoint __call(NBL_CONST_REF_ARG(FloatingPoint) x)
+	{
+		// TODO: no way this is optimal, find a better implementation
+		float tmp;
+		if (std::abs(std::modf(x, &tmp)) == 0.5f)
+		{
+			int32_t result = static_cast<int32_t>(x);
+			if (result % 2 != 0)
+				result >= 0 ? ++result : --result;
+			return result;
+		}
+
+		return std::round(x);
+	}
+};
+
+template<typename FloatingPoint>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointScalar<FloatingPoint>)
+struct fma_helper<FloatingPoint NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<FloatingPoint>) >
+{
+	static FloatingPoint __call(NBL_CONST_REF_ARG(FloatingPoint) x, NBL_CONST_REF_ARG(FloatingPoint) y, NBL_CONST_REF_ARG(FloatingPoint) z)
+	{
+		return std::fma(x, y, z);
+	}
+};
+
+template<typename T, typename U>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointScalar<T> && concepts::IntegralScalar<U>)
+struct ldexp_helper<T, U NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<T> && concepts::IntegralScalar<U>) >
+{
+	static T __call(NBL_CONST_REF_ARG(T) arg, NBL_CONST_REF_ARG(U) exp)
+	{
+		return std::ldexp(arg, exp);
+	}
+};
+
+template<typename T>
+requires concepts::FloatingPointScalar<T>
+struct modfStruct_helper<T>
+{
+	using return_t = ModfOutput<T>;
+	static inline return_t __call(const T val)
+	{
+		return_t output;
+		output.fractionalPart = std::modf(val, &output.wholeNumberPart);
+
+		return output;
+	}
+};
+
+template<typename T>
+requires concepts::FloatingPointScalar<T>
+struct frexpStruct_helper<T>
+{
+	using return_t = FrexpOutput<T>;
+	static inline return_t __call(const T val)
+	{
+		return_t output;
+		output.significand = std::frexp(val, &output.exponent);
+
+		return output;
+	}
+};
+
+#endif // C++ only specializations
+
+// C++ and HLSL specializations
+
+template<typename FloatingPoint>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointScalar<FloatingPoint>)
+struct erfInv_helper<FloatingPoint NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<FloatingPoint>) >
+{
+	static FloatingPoint __call(NBL_CONST_REF_ARG(FloatingPoint) _x)
+	{
+		FloatingPoint x = clamp<FloatingPoint>(_x, -0.99999, 0.99999);
+
+		FloatingPoint w = -log_helper<FloatingPoint>::__call((1.0 - x) * (1.0 + x));
+		FloatingPoint p;
+		if (w < 5.0)
+		{
+			w -= 2.5;
+			p = 2.81022636e-08;
+			p = 3.43273939e-07 + p * w;
+			p = -3.5233877e-06 + p * w;
+			p = -4.39150654e-06 + p * w;
+			p = 0.00021858087 + p * w;
+			p = -0.00125372503 + p * w;
+			p = -0.00417768164 + p * w;
+			p = 0.246640727 + p * w;
+			p = 1.50140941 + p * w;
+		}
+		else
+		{
+			w = sqrt_helper<FloatingPoint>::__call(w) - 3.0;
+			p = -0.000200214257;
+			p = 0.000100950558 + p * w;
+			p = 0.00134934322 + p * w;
+			p = -0.00367342844 + p * w;
+			p = 0.00573950773 + p * w;
+			p = -0.0076224613 + p * w;
+			p = 0.00943887047 + p * w;
+			p = 1.00167406 + p * w;
+			p = 2.83297682 + p * w;
+		}
+		return p * x;
+	}
+};
```

Comment:
why only scalar?

Reply:
also due to how HLSL and C++ work (disagreement) and lack of support for C++23 float literal size suffices, every single float literal needs to be wrapped in `float_t`, e.g. `float_t(0.3839898298298)`

Because C++ will take `2.8` to be a double, and HLSL will make it a single when in-fact you would have wanted a double

Just one of those annoying things

Reply:
Also the way you wrote this, in C++ when you do an op involving a double, other operand gets promoted to double IIRC (at least the result does), so when `is_same_v<FloatingPoint,float32_t>` you'd end up with some ops being done as fp32 then their result getting promoted to f64, some ops done in f64 only to get truncated back down to 32

Reply:
> why only scalar?

i specialize for vectors below https://github.com/Devsh-Graphics-Programming/Nabla/blob/new_tgmath/include/nbl/builtin/hlsl/tgmath/impl.hlsl#L437